src/memory.c

```diff
@@ -56,6 +71,19 @@ char *jsonp_strndup(const char *str, size_t len)
 
 void json_set_alloc_funcs(json_malloc_t malloc_fn, json_free_t free_fn)
 {
-    do_malloc = malloc_fn;
-    do_free = free_fn;
+    do_malloc.normal = malloc_fn;
+    do_free.normal = free_fn;
+    user_arg = NULL; 
 }
+
+void json_set_alloc_funcs_arg(json_malloc_arg_t malloc_fn, json_free_arg_t free_fn, void *arg)
+{
+    if (arg == NULL) {
+        json_set_alloc_funcs((json_malloc_t) malloc_fn, (json_free_t) free_fn);
```

Comment:
Calling `json_set_alloc_funcs_arg()` with a `NULL` arg is most likely a mistake. If the passed malloc/free functions do not actually take an extra arg, then the user calling this variant would cause compiler warnings. If the passed malloc/free functions do take an extra arg and make it this far without compiler warnings, then calling them without the extra arg causes unexpected values from the stack or registers to be interpreted as the extra arg. (I think that for many architectures / calling-conventions, it works fine to call with an extra `NULL` argument that is not expected, it will just be ignored - but it's a mess either way.)

Reply:
My preference is to either assert or have separate storage for the `normal` vs `with_arg` functions.  If someone uses `json_set_alloc_funcs_arg` then we must not call the passed function without the arg.  For separate function storage I think `json_set_alloc_funcs_arg` would need to NULL out the `normal` functions and `json_set_alloc_funcs` would need to NULL out the `with_arg` functions so only the last functions set could take effect.